### PR TITLE
RxJava improvements

### PR DESCRIPTION
**Pull request by vietj**
- deprecate `*Observable` methods with corresponding `rx*` methods that returns an RxJava `Single` that calls the method at subscription and provide closer to RxJava interactions
- method having a `ReadStream<T>` argument are overloaded with an `Observable<T>` version
- replace the return type of the `Object getDelegate()` methods to be the actual delegate type as a raw type

### CODE REVIEW

**Comment by vietj** [comment]
please review @pmlopes @jponge @pmlopes @tsegismont or anyone else

**Comment by dano** [comment]
The `getDelegate()` change is appreciated. 

We're looking at a nice, long deprecation window for the `*Observable()` methods, right? It'll be a fairly large effort to convert those to the `rx*` equivalents in my codebase.

**Comment by vietj** [comment]
@dano yes it should remain there for a couple of versions at least

**Comment by vietj** [comment]
@dano if you foresee other improvements we could do to RxJava, don't hesitate to discuss it on vertx-dev, for 3.4 we really want to improve the RxJava usability. If you are using Vert.x HttpClient, you want also to look at the new Web Client that plays quite nicely with this PR (actually it drove it)

**Comment by dano** [comment]
@vietj Sure, I appreciate the invitation.

One comment on this PR: it seems like there should probably be `SingleHandler`/`SingleFuture` classes, in addition to `ObservableHandler`/`ObservableFuture`, since `Handler<AsyncResult<T>>` will now be turned into a `Single<T>`.

File: rx-java/src/main/java/io/vertx/rx/java/ReadStreamSubscriber.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    action.run();
+  }
+
+  @Override
+  public ReadStream<J> endHandler(Handler<Void> handler) {
+    synchronized (this) {
+      if (completed == null || pending.size() > 0) {
+        endHandler = handler;
+      } else {
+        if (handler != null) {
```
Comment (tsegismont) [inline rx-java/src/main/java/io/vertx/rx/java/ReadStreamSubscriber.java]: Is there a reason to not throw the ISE if the provided handler is null?

File: rx-java/src/main/java/io/vertx/rx/java/ReadStreamSubscriber.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    return this;
+  }
+
+  @Override
+  public ReadStream<J> exceptionHandler(Handler<Throwable> handler) {
+    synchronized (this) {
+      if (completed == null || pending.size() > 0) {
+        exceptionHandler = handler;
+      } else {
+        if (handler != null) {
```
Comment (tsegismont) [inline rx-java/src/main/java/io/vertx/rx/java/ReadStreamSubscriber.java]: Same question as for the end handler

File: rx-java/src/test/java/io/vertx/rx/java/test/ReadStreamSubscriberTest.java [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    private ReadStreamSubscriber<String, String> subscriber = new ReadStreamSubscriber<>(Function.identity());
+    private long requested;
+    private int seq;
+
+    public Sender() {
+      subscriber.setProducer(n -> requested += n);
+    }
+
+    void emit() {
+      subscriber.onNext("" + seq++);
```
Comment (tsegismont) [inline rx-java/src/test/java/io/vertx/rx/java/test/ReadStreamSubscriberTest.java]: `String.valueOf(seq++)` ?

File: rx-java/src/test/java/io/vertx/rx/java/test/ReadStreamSubscriberTest.java [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+    Receiver receiver = new Receiver();
+    receiver.subscribe(sender.subscriber);
+    sender.emit(3);
+    sender.complete();
+    sender.subscriber.resume();
+    receiver.assertItems("0", "1", "2").assertEnded();
+  }
+
+  @Test
+  public void testSetNullHandlersInEndHandler() {
```
Comment (tsegismont) [inline rx-java/src/test/java/io/vertx/rx/java/test/ReadStreamSubscriberTest.java]: Sorry I don't get the purpose of this one... :(

File: rx-java/src/main/resources/vertx-rxjava/template/common.templ [hunk 5]
```diff
@@ -171,6 +183,90 @@
...
+      futParams,
+      method.comment,
+      method.doc,
+      method.staticMethod,
+      method.defaultMethod,
+      method.typeParams);
+    return futMethod;
+  }
+
+  def hasReadStream(method) {
```
Comment (vietj) [inline rx-java/src/main/resources/vertx-rxjava/template/common.templ]: I think it's a leftover of previous attemps

File: rx-java/src/test/java/io/vertx/rx/java/test/ReadStreamSubscriberTest.java (hunk 3 above)
Comment (vietj) [inline rx-java/src/test/java/io/vertx/rx/java/test/ReadStreamSubscriberTest.java]: Less readable :-)

**Comment by tsegismont** [comment]
Mater of taste :)

2016-12-12 7:55 GMT+01:00 Julien Viet <notifications@github.com>:

File: rx-java/src/main/java/io/vertx/rx/java/ReadStreamSubscriber.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+public class ReadStreamSubscriber<R, J> extends Subscriber<R> implements ReadStream<J> {
+
+  private static final Runnable NOOP_ACTION = () -> {};
+  private static final Throwable DONE_SENTINEL = new Throwable();
+
+  public static final int BUFFER_SIZE = 16;
+
+  public static <R, J> ReadStream<J> asReadStream(Observable<R> observable, Function<R, J> adapter) {
```
Comment (tsegismont) [inline rx-java/src/main/java/io/vertx/rx/java/ReadStreamSubscriber.java]: And what would the adapter function be? Won't it be identity again? Do we want to convert the buffer in HttpRequest#send ? How does codegen know how to convert?

File: rx-java/src/main/java/io/vertx/rx/java/ReadStreamSubscriber.java (hunk 1 above)
Comment (tsegismont) [inline rx-java/src/main/java/io/vertx/rx/java/ReadStreamSubscriber.java]: I guess the logic is that null is a valid argument even when the stream is completed and there are no pending buffers.

**Comment by tsegismont** [comment]
Perfect, thanks for the details.

2016-12-12 7:57 GMT+01:00 Julien Viet <notifications@github.com>:

**Pull request merged**